src/ncp/ncp_base_mtd.cpp

```diff
@@ -818,6 +823,139 @@ template <> otError NcpBase::HandlePropertyRemove<SPINEL_PROP_THREAD_ON_MESH_NET
 }
 #endif // OPENTHREAD_ENABLE_BORDER_ROUTER
 
+#if OPENTHREAD_ENABLE_SERVICE
+
+template <> otError NcpBase::HandlePropertyGet<SPINEL_PROP_SERVER_ALLOW_LOCAL_DATA_CHANGE>(void)
+{
+    return mEncoder.WriteBool(mAllowLocalServerDataChange);
+}
+
+template <> otError NcpBase::HandlePropertySet<SPINEL_PROP_SERVER_ALLOW_LOCAL_DATA_CHANGE>(void)
+{
+    bool    value                    = false;
+    otError error                    = OT_ERROR_NONE;
+    bool    shouldRegisterWithLeader = false;
+
+    SuccessOrExit(error = mDecoder.ReadBool(value));
+
+    // Register any server data changes on transition from `true` to `false`.
+    shouldRegisterWithLeader = (mAllowLocalServerDataChange == true) && (value == false);
+
+    mAllowLocalServerDataChange = value;
+
+exit:
+
+    if (shouldRegisterWithLeader)
+    {
+        otServerRegister(mInstance);
+    }
+
+    return error;
+}
+
+template <> otError NcpBase::HandlePropertyInsert<SPINEL_PROP_SERVER_SERVICES>(void)
+{
+    otError         error = OT_ERROR_NONE;
+    otServiceConfig cfg;
+    bool            stable;
+    const uint8_t * data;
+    uint16_t        dataLen;
+
+    VerifyOrExit(mAllowLocalServerDataChange == true, error = OT_ERROR_INVALID_STATE);
+
+    SuccessOrExit(error = mDecoder.ReadUint32(cfg.mEnterpriseNumber));
+    SuccessOrExit(error = mDecoder.ReadDataWithLen(data, dataLen));
+
+    VerifyOrExit((dataLen <= sizeof(cfg.mServiceData) && dataLen <= UINT8_MAX), error = OT_ERROR_INVALID_ARGS);
```

Comment:
Do we need the `dataLen <= UINT8_MAX`? Isn't it covered by stricter `dataLen <= sizeof(cfg.mServiceData)`?

Reply:
It is, but I figured that if somebody changes the size of `cfg.mServiceData` in the future to be greater than 255, and changed the length parameter to a wider type (e.g. uint16_t), then unknowingly to the author it'd be possible to specify a length that overflows because of the narrowing cast on line 872: `static_cast<uint8_t>(dataLen)`.

Perhaps it'd be better to split this out in to static assert, something like:
```
VerifyOrExit((dataLen <= sizeof(cfg.mServiceData), error = OT_ERROR_INVALID_ARGS);
OT_STATIC_ASSERT(sizeof(cfg.mServiceData) <= UINT8_MAX, "Invalid cast to uint8");
```

Reply:
I am fine keeping it as is.